src/liger_kernel/transformers/model/llama.py

```diff
@@ -247,3 +253,47 @@ def lce_forward(
         hidden_states=outputs.hidden_states,
         attentions=outputs.attentions,
     )
+
+
+def lce_maybe_trainable_lm_head(self, hidden_states, hidden_size, labels, **loss_kwargs):
+    lm_head = self.lm_head
+
+    # Unwrap the module if lm_head has been added as trainable module in PEFT LoRA configuration,
+    # i.e. listed in the modules_to_save field of LoraConfig, so the lm_head weights are read
+    # from the unwrapped module.
+    # See https://huggingface.co/docs/peft/package_reference/lora for reference.
+    if PEFT_AVAILABLE and isinstance(lm_head, ModulesToSaveWrapper):
+        lm_head = lm_head.modules_to_save.default
+
+    # If FSDP is used and lm_head is trainable, e.g., during full fine-tuning or with LoRA,
+    # reading the lm_head module weights and calling the kernel must be done within FSDP forward pass
+    # so the module entire parameters are summoned and kept in memory during the kernel execution.
+    if isinstance(lm_head, FullyShardedDataParallel):
```

Comment:
I wonder if this is sufficient to fix the issue -- in majority of the cases, `lm_head` on its own won't be wrapped into a leaf FSDP module. Rather people usually wrap the entire `AutoModelForCausalLM` as root FSDP, and each decoder layer in it as leaf FSDP, leaving input embedding and `lm_head` just standard `nn.Module` managed by the root FSDP instance; That being said `lm_head` is (almost) never going to be an FSDP instance except a user explicitly configure the wrapping policy to wrap `lm_head` into FSDP in separate 🤔